File: examples/cpp/4file_custom_pph_using_opencv.cpp
/*
The MIT License

Copyright (c) 2024, Prominence AI, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in-
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

/*
#
# This example demonstrates the use of a Frame-Capture Sink to encode and
# save video frames to JPEG files on client/viewer demand.
#
# An ODE Frame-Capture Action is provided to The Frame-Capture Sink on creation.
# A client "capture_complete_listener" is added to the the Action to be notified
# when each new file is saved (the ODE Action performs the actual frame-capture).
#
# Child Players (to play the captured image) and Mailers (to mail the image) can
# be added to the ODE Frame-Capture action as well (not shown).
#
# A Custom Pad Probe Handler (PPH) is added to the sink-pad of the OSD component
# to process every buffer flowing over the pad by:
#    - Retrieving the batch-metadata and its list of frame metadata structures
#      (only one frame per batched-buffer with 1 Source)
#    - Retrieving the list of object metadata structures from the frame metadata.
#    - Iterating through the list of objects looking for the first occurrence of
#      a bicycle. 
#    - If detected, the current frame-number is schedule to be captured by the
#      Frame-Capture Sink using its Frame-Capture Action.
#
#          dsl_sink_frame_capture_schedule('frame-capture-sink', 
#                   frame_meta.frame_num)
#
# Note: The Custom PPH will schedule every frame with a bicycle to be captured!
#
# IMPORT All captured frames are copied and buffered in the Sink's processing
# thread. The encoding and saving of each buffered frame is done in the 
# g-idle-thread, therefore, the capture-complete notification is asynchronous.
#
*/

#include <iostream>
#include <glib.h>
#include <gst/gst.h>
#include <gstnvdsmeta.h>

#include "opencv2/imgproc/imgproc.hpp"
#include "opencv2/highgui/highgui.hpp"

#include "DslApi.h"

// Use the DSL Surface Transform utility class
#include "DslSurfaceTransform.h"

uint device_type;
uint unique_id=0;

std::wstring uri_h265(
    L"/opt/nvidia/deepstream/deepstream/samples/streams/sample_1080p_h265.mp4");

// Config and model-engine files - Jetson and dGPU
std::wstring primary_infer_config_file(
    L"/opt/nvidia/deepstream/deepstream/samples/configs/deepstream-app/config_infer_primary.txt");
std::wstring primary_model_engine_file(
    L"/opt/nvidia/deepstream/deepstream/samples/models/Primary_Detector/resnet18_trafficcamnet.etlt_b8_gpu0_int8.engine");

// Config file used by the IOU Tracker    
std::wstring iou_tracker_config_file(
    L"/opt/nvidia/deepstream/deepstream/samples/configs/deepstream-app/config_tracker_IOU.yml");

// Same dimensions as Pipeline Streammuxer 
int TILER_WIDTH = DSL_1K_HD_WIDTH;
int TILER_HEIGHT = DSL_1K_HD_HEIGHT;

uint WINDOW_WIDTH = 1280;
uint WINDOW_HEIGHT = 720;

// 
// Function to be called on XWindow KeyRelease event
// 
void xwindow_key_event_handler(const wchar_t* in_key, void* client_data)
{   
    std::wstring wkey(in_key); 
    std::string key(wkey.begin(), wkey.end());
    std::cout << "key released = " << key << std::endl;
    key = std::toupper(key[0]);
    if(key == "P"){
        dsl_pipeline_pause(L"pipeline");
    } else if (key == "R"){
        dsl_pipeline_play(L"pipeline");
    } else if (key == "Q" or key == "" or key == ""){
        std::cout << "Main Loop Quit" << std::endl;
        dsl_pipeline_stop(L"pipeline");
        dsl_main_loop_quit();
    }
}

// 
// Function to be called on XWindow Delete event
//
void xwindow_delete_event_handler(void* client_data)
{
    std::cout << "delete window event" <<std::endl;
    dsl_pipeline_stop(L"pipeline");
    dsl_main_loop_quit();
}
    
// 
// Function to be called on End-of-Stream (EOS) event
// 
void eos_event_listener(void* client_data)
{
    std::cout <<"Pipeline EOS event" <<std::endl;
    dsl_pipeline_stop(L"pipeline");
    dsl_main_loop_quit();
}    

// 
// Function to be called on every change of Pipeline state
// 
void state_change_listener(uint old_state, uint new_state, void* client_data)
{
    std::cout << "previous state = " << dsl_state_value_to_string(old_state) 
        << ", new state = " << dsl_state_value_to_string(new_state) << std::endl;
}

// 
// Custom Pad Probe Handler function called with every buffer
// 
uint custom_pad_probe_handler(void* buffer, void* user_data)
{
    // Retrieve batch metadata from the gst_buffer
    NvDsBatchMeta* pBatchMeta = gst_buffer_get_nvds_batch_meta((GstBuffer*)buffer);
    
    // For each frame in the batched meta data
    for (NvDsMetaList* pFrameMetaList = pBatchMeta->frame_meta_list; 
        pFrameMetaList; pFrameMetaList = pFrameMetaList->next)
    {
        // Check for valid frame data
        NvDsFrameMeta* pFrameMeta = (NvDsFrameMeta*)(pFrameMetaList->data);
        if (pFrameMeta != NULL)
        {
            
            // Map the current buffer
            std::unique_ptr<DSL::DslMappedBuffer> pMappedBuffer = 
                std::unique_ptr<DSL::DslMappedBuffer>(new DSL::DslMappedBuffer(
                    (GstBuffer*)buffer));
                
            NvBufSurfaceMemType transformMemType = (
                device_type == DSL_GPU_TYPE_INTEGRATED)
                ? NVBUF_MEM_DEFAULT
                : NVBUF_MEM_CUDA_PINNED;
        
            // Transforming only one frame in the batch, so create a copy of the single 
            // surface ... becoming our new source surface. This creates a new mono 
            // (non-batched) surface copied from the "batched frames" using the batch id 
            // as the index
            DSL::DslMonoSurface monoSurface(pMappedBuffer->pSurface, 
                pFrameMeta->batch_id);

            // Coordinates and dimensions for our destination surface.
            int left(0), top(0);

            // capturing full frame
            int width = pMappedBuffer->GetWidth(pFrameMeta->batch_id);
            int height = pMappedBuffer->GetHeight(pFrameMeta->batch_id);
            
            // New "create params" for our destination surface. we only need one 
            // surface so set memory allocation (for the array of surfaces) size to 0
            DSL::DslSurfaceCreateParams surfaceCreateParams(monoSurface.gpuId, 
                width, height, 0, NVBUF_COLOR_FORMAT_RGBA, transformMemType);
            
            // New Destination surface with a batch size of 1 for transforming 
            // the single surface. Use batch-id as unique id. 
            std::unique_ptr<DSL::DslBufferSurface> pBufferSurface = 
                std::unique_ptr<DSL::DslBufferSurface>(new DSL::DslBufferSurface(1, 
                surfaceCreateParams, pFrameMeta->batch_id));

            // New "transform params" for the surface transform, croping
            DSL::DslTransformParams transformParams(left, top, width, height);
            
            // New "Cuda stream" for the surface transform
            DSL::DslCudaStream dslCudaStream(monoSurface.gpuId);
            
            // New "Transform Session" config params using the new Cuda stream
            DSL::DslSurfaceTransformSessionParams dslTransformSessionParams(
                monoSurface.gpuId, dslCudaStream);
            
            // Set the "Transform Params" for the current tranform session
            if (!dslTransformSessionParams.Set())
            {
                std::cout << "failed to set Transform Params" << std::endl;
                return DSL_PAD_PROBE_REMOVE;
            }
            
            // We can now transform our Mono Source surface to the first (and only) 
            // surface in the batched buffer.
            if (!pBufferSurface->TransformMonoSurface(monoSurface, 0, transformParams))
            {
                std::cout << "failed to transform Mono Surface" << std::endl;
                return DSL_PAD_PROBE_REMOVE;
            }

            // Map the tranformed surface for read
            if (!pBufferSurface->Map())
            {
                std::cout << "failed to map the transformed buffer" << std::endl;
                return DSL_PAD_PROBE_REMOVE;
            }

            std::cout << "New Buffer Surface mapped succesfully" << std::endl;

            // New background Mat for our image
            cv::Mat bgrFrame = cv::Mat(cv::Size(width, height), CV_8UC3);

            // Use openCV to remove padding
            cv::Mat inMat = cv::Mat(height, width, CV_8UC4, 
                (&(*pBufferSurface))->surfaceList[0].mappedAddr.addr[0],
                (&(*pBufferSurface))->surfaceList[0].pitch);

            cv::cvtColor (inMat, bgrFrame, cv::COLOR_RGBA2BGR);
            
            // Do something with the cv::mat - the below saves the mat to jpeg file

            // Generate the image file name from the date-time string
            std::ostringstream fileNameStream;
            fileNameStream << "batch_" 
                << std::setw(2) << std::setfill('0') << pBufferSurface->GetUniqueId()
                << "_" << pBufferSurface->GetDateTimeStr() << ".jpeg";
                
            // Generate the filespec from the output dir and file name
            std::string filespec = "./" + 
                fileNameStream.str();
            cv::imwrite(filespec.c_str(), bgrFrame);
            
        }
    }
    return DSL_PAD_PROBE_OK;
}

int main(int argc, char** argv)
{
    DslReturnType retval = DSL_RESULT_FAILURE;

    // Since we're not using args, we can Let DSL initialize GST on first call    
    while(true) 
    {    
        // Get the Device type - DSL_GPU_TYPE_INTEGRATED | DSL_GPU_TYPE_DISCRETE
        device_type = dsl_info_gpu_type_get(0);

        // 4 New File Sources
        retval = dsl_source_file_new(L"file-source-0", uri_h265.c_str(), true);
        if (retval != DSL_RESULT_SUCCESS) break;
        retval = dsl_source_file_new(L"file-source-1", uri_h265.c_str(), true);
        if (retval != DSL_RESULT_SUCCESS) break;
        retval = dsl_source_file_new(L"file-source-2", uri_h265.c_str(), true);
        if (retval != DSL_RESULT_SUCCESS) break;
        retval = dsl_source_file_new(L"file-source-3", uri_h265.c_str(), true);
        if (retval != DSL_RESULT_SUCCESS) break;

        // lower the frame-rates - this trivial example saves every frame to jpeg.
        retval = dsl_source_video_buffer_out_frame_rate_set(L"file-source-0", 1, 1);
        if (retval != DSL_RESULT_SUCCESS) break;
        retval = dsl_source_video_buffer_out_frame_rate_set(L"file-source-1", 1, 1);
        if (retval != DSL_RESULT_SUCCESS) break;
        retval = dsl_source_video_buffer_out_frame_rate_set(L"file-source-2", 1, 1);
        if (retval != DSL_RESULT_SUCCESS) break;
        retval = dsl_source_video_buffer_out_frame_rate_set(L"file-source-3", 1, 1);
        if (retval != DSL_RESULT_SUCCESS) break;

        // New Primary GIE using the filespecs defined above, with interval = 4
        retval = dsl_infer_gie_primary_new(L"primary-gie", 
            primary_infer_config_file.c_str(), 
            primary_model_engine_file.c_str(), 4);
        if (retval != DSL_RESULT_SUCCESS) break;
        
        // New IOU Tracker, setting operational width and height of input frame
        retval = dsl_tracker_new(L"iou-tracker", 
            iou_tracker_config_file.c_str(), 480, 272);
        if (retval != DSL_RESULT_SUCCESS) break;

        // New Tiler, setting width and height
        retval = dsl_tiler_new(L"tiler", TILER_WIDTH, TILER_HEIGHT);
        if (retval != DSL_RESULT_SUCCESS) break;

        // New Custom Pad Probe Handler (PPH) to call our handler function above
        retval = dsl_pph_custom_new(L"custom-pph", custom_pad_probe_handler, NULL);
        if (retval != DSL_RESULT_SUCCESS) break;

        // Add the Custom PPH to the sink (input) pad of the Tiler component.
        retval = dsl_tiler_pph_add(L"tiler", L"custom-pph", DSL_PAD_SINK);
        if (retval != DSL_RESULT_SUCCESS) break;

        // New OSD with text, clock and bbox display all enabled. 
        retval = dsl_osd_new(L"on-screen-display", true, true, true, false);
        if (retval != DSL_RESULT_SUCCESS) break;

        // New Window Sink, 0 x/y offsets.
        retval = dsl_sink_window_egl_new(L"egl-sink", 0, 0, 
            WINDOW_WIDTH, WINDOW_HEIGHT);
        if (retval != DSL_RESULT_SUCCESS) break;
    
        // Add the XWindow event handler functions defined above
        retval = dsl_sink_window_key_event_handler_add(L"egl-sink", 
            xwindow_key_event_handler, NULL);
        if (retval != DSL_RESULT_SUCCESS) break;

        retval = dsl_sink_window_delete_event_handler_add(L"egl-sink", 
            xwindow_delete_event_handler, NULL);
        if (retval != DSL_RESULT_SUCCESS) break;

        // Create a list of Pipeline Components to add to the new Pipeline.
        const wchar_t* components[] = {
            L"file-source-0", L"file-source-1", L"file-source-2", L"file-source-3",
            L"primary-gie", L"iou-tracker", L"tiler", L"on-screen-display", 
            L"egl-sink", NULL};
        
        // Add all the components to our pipeline
        retval = dsl_pipeline_new_component_add_many(L"pipeline", components);
        if (retval != DSL_RESULT_SUCCESS) break;
            
        // Add the EOS listener function defined above
        retval = dsl_pipeline_eos_listener_add(L"pipeline", eos_event_listener, NULL);
        if (retval != DSL_RESULT_SUCCESS) break;

        // Play the pipeline
        retval = dsl_pipeline_play(L"pipeline");
        if (retval != DSL_RESULT_SUCCESS) break;

        // Start and join the main-loop
        dsl_main_loop_run();
        break;
    }
    
    // Print out the final result
    std::wcout << dsl_return_value_to_string(retval) << std::endl;

    dsl_delete_all();

    std::cout<<"Goodbye!"<<std::endl;  
    return 0;
}
            

File: examples/python/4file_custom_pph_using_opencv.py
################################################################################
# The MIT License
#
# Copyright (c) 2019-2024, Prominence AI, Inc.
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.
################################################################################

################################################################################
#
# This simple example demonstrates how to use OpenCV with NVIDIA's pyds.
# The Pipeline used in this example is built with :
#   - 4 URI Sources
#   - Primary GST Inference Engine (PGIE)
#   - IOU Tracker
#   - Tiler
#   - On-Screen Display (OSD)
#   - Window Sink
# 
# A Custom Pad-Probe-Handler is added to the Sink-Pad of the Tiler
# to process the frame meta-data for each buffer received. The handler
# demonstrates how to 
#   - use pyds.get_nvds_buf_surface() to get a buffer surface.
#   - convert a frame to numpy array format with np.array().
#   - convert the array into cv2 default BGRA format using cv2.cvtColor().
#   - save the array as an image using opencv cv2.imwrite().
#
# IMPORTANT! pyds.get_nvds_buf_surface() requires 
#   1. The color format of the buffer must be set to RGBA by calling
#      dsl_source_video_buffer_out_format_set()
#   2. The memory type must be set to DSL_NVBUF_MEM_TYPE_CUDA_UNIFIED
#      if running on dGPU. This is done by calling 
#        * dsl_pipeline_streammux_nvbuf_mem_type_set() - if using old streammux
#        * dsl_component_nvbuf_mem_type_set_many() - with all sources if using 
#          new streammux. 
#     
# IMPORTANT! The output folders (1 per source) must be created first
#   ./stream_0, ./stream_1, ./stream_2, ./stream_3,
# 
# The example registers handler callback functions with the Pipeline for:
#   - key-release events
#   - delete-window events
#   - end-of-stream EOS events
#   - Pipeline change-of-state events
#  
################################################################################

#!/usr/bin/env python

import sys
from dsl import *
import pyds
import numpy as np
import cv2

source_uri = "/opt/nvidia/deepstream/deepstream/samples/streams/sample_1080p_h264.mp4"


# Filespecs (Jetson and dGPU) for the Primary GIE
primary_infer_config_file = \
    '/opt/nvidia/deepstream/deepstream/samples/configs/deepstream-app/config_infer_primary.txt'
primary_model_engine_file = \
    '/opt/nvidia/deepstream/deepstream/samples/models/Primary_Detector/resnet18_trafficcamnet.etlt_b8_gpu0_int8.engine'

# Filespec for the IOU Tracker config file
iou_tracker_config_file = \
    '/opt/nvidia/deepstream/deepstream/samples/configs/deepstream-app/config_tracker_IOU.yml'

WINDOW_WIDTH = 1280
WINDOW_HEIGHT = 720

## 
# Function to be called on XWindow KeyRelease event
## 
def xwindow_key_event_handler(key_string, client_data):
    print('key released = ', key_string)
    if key_string.upper() == 'P':
        dsl_pipeline_pause('pipeline')
    elif key_string.upper() == 'R':
        dsl_pipeline_play('pipeline')
    elif key_string.upper() == 'Q' or key_string == '' or key_string == '':
        dsl_pipeline_stop('pipeline')
        dsl_main_loop_quit()
 
## 
# Function to be called on XWindow Delete event
## 
def xwindow_delete_event_handler(client_data):
    print('delete window event')
    dsl_pipeline_stop('pipeline')
    dsl_main_loop_quit()

## 
# Function to be called on End-of-Stream (EOS) event
## 
def eos_event_listener(client_data):
    print('Pipeline EOS event')
    dsl_pipeline_stop('pipeline')
    dsl_main_loop_quit()

## 
# Function to be called on every change of Pipeline state
## 
def state_change_listener(old_state, new_state, client_data):
    print('previous state = ', old_state, ', new state = ', new_state)
    if new_state == DSL_STATE_PLAYING:
        dsl_pipeline_dump_to_dot('pipeline', "state-playing")

## 
# Custom PPH added to the sink-pad (input) of the Tiler
## 
def custom_pad_probe_handler(buffer, user_data):

    # Retrieve batch metadata from the gst_buffer
    # IMPORTANT! do not use the hash function to cast the buffer.
    
    batch_meta = pyds.gst_buffer_get_nvds_batch_meta(buffer)
    l_frame = batch_meta.frame_meta_list
    while l_frame is not None:
        try:
            # Note that l_frame.data needs a cast to pyds.NvDsFrameMeta
            # The casting is done by pyds.glist_get_nvds_frame_meta()
            # The casting also keeps ownership of the underlying memory
            # in the C code, so the Python garbage collector will leave
            # it alone.
            frame_meta = pyds.glist_get_nvds_frame_meta(l_frame.data)
        except StopIteration:
            break

        # Saving every 100th frame for each source.  
        if frame_meta.frame_num%100 == 0:

            n_frame = pyds.get_nvds_buf_surface(buffer, frame_meta.batch_id)

            # convert the python array into numpy array format.
            frame_image = np.array(n_frame,copy=True,order='C')

            # covert the array into cv2 default BGRA format
            frame_image = cv2.cvtColor(frame_image,cv2.COLOR_RGBA2BGRA)

            filename = "./stream_"+str(frame_meta.pad_index) + \
                "/frame_"+str(frame_meta.frame_num)+".jpg"
            print(filename)

            # write out the image
            cv2.imwrite(filename,frame_image)

        try:
            l_frame=l_frame.next
        except StopIteration:
            break
    return DSL_PAD_PROBE_OK

def main(args):

    # Since we're not using args, we can Let DSL initialize GST on first call
    while True:

        # 4 New URI File Sources using the filespec defined above
        retval = dsl_source_uri_new('uri-source-0', source_uri, False, False, 0)
        if retval != DSL_RETURN_SUCCESS:
            break

        retval = dsl_source_uri_new('uri-source-1', source_uri, False, False, 0)
        if retval != DSL_RETURN_SUCCESS:
            break

        retval = dsl_source_uri_new('uri-source-2', source_uri, False, False, 0)
        if retval != DSL_RETURN_SUCCESS:
            break

        retval = dsl_source_uri_new('uri-source-3', source_uri, False, False, 0)
        if retval != DSL_RETURN_SUCCESS:
            break

        # IMPORTANT! We must set the buffer format to RGBA for each source.        
        retval = dsl_source_video_buffer_out_format_set('uri-source-0', 
            DSL_VIDEO_FORMAT_RGBA)
        if retval != DSL_RETURN_SUCCESS:
            break
            
        retval = dsl_source_video_buffer_out_format_set('uri-source-1', 
            DSL_VIDEO_FORMAT_RGBA)
        if retval != DSL_RETURN_SUCCESS:
            break
            
        retval = dsl_source_video_buffer_out_format_set('uri-source-2', 
            DSL_VIDEO_FORMAT_RGBA)
        if retval != DSL_RETURN_SUCCESS:
            break
            
        retval = dsl_source_video_buffer_out_format_set('uri-source-3', 
            DSL_VIDEO_FORMAT_RGBA)
        if retval != DSL_RETURN_SUCCESS:
            break
            
           
        # New Primary GIE using the filespecs above with interval = 0
        retval = dsl_infer_gie_primary_new('primary-gie', 
            primary_infer_config_file, primary_model_engine_file, 0)
        if retval != DSL_RETURN_SUCCESS:
            break

        # New IOU Tracker, setting operational width and hieght
        retval = dsl_tracker_new('iou-tracker', iou_tracker_config_file, 480, 272)
        if retval != DSL_RETURN_SUCCESS:
            break

        # New Tiler with dimensions for two tiles - for the two sources
        retval = dsl_tiler_new('tiler', WINDOW_WIDTH, WINDOW_HEIGHT)
        if retval != DSL_RETURN_SUCCESS:
            break

        # New Custom Pad Probe Handler to call Nvidia's example callback 
        # for handling the Batched Meta Data
        retval = dsl_pph_custom_new('custom-pph', 
            client_handler=custom_pad_probe_handler, client_data=None)

        # Add the custom PPH to the Sink pad (input) of the Tiler.
        retval = dsl_tiler_pph_add('tiler', 
            handler='custom-pph', pad=DSL_PAD_SINK)
        if retval != DSL_RETURN_SUCCESS:
            break
                # New OSD with text, clock and bbox display all enabled. 
        retval = dsl_osd_new('on-screen-display', 
            text_enabled=True, clock_enabled=True, 
            bbox_enabled=True, mask_enabled=False)
        if retval != DSL_RETURN_SUCCESS:
            break

        # New Window Sink, 0 x/y offsets and dimensions defined above.
        retval = dsl_sink_window_egl_new('egl-sink', 0, 0, 
            WINDOW_WIDTH, WINDOW_HEIGHT)
        if retval != DSL_RETURN_SUCCESS:
            break

        # Add the XWindow event handler functions defined above
        retval = dsl_sink_window_key_event_handler_add('egl-sink', 
            xwindow_key_event_handler, None)
        if retval != DSL_RETURN_SUCCESS:
            break
        retval = dsl_sink_window_delete_event_handler_add('egl-sink', 
            xwindow_delete_event_handler, None)
        if retval != DSL_RETURN_SUCCESS:
            break

        # Add all the components to our pipeline
        retval = dsl_pipeline_new_component_add_many('pipeline', 
            ['uri-source-0', 'uri-source-1', 'uri-source-2', 'uri-source-3', 
             'primary-gie', 'iou-tracker',  'tiler', 'on-screen-display', 
             'egl-sink', None])
        if retval != DSL_RETURN_SUCCESS:
            break

        # IMPORTANT! pyds.get_nvds_buf_surface requires the buffer memory to use
        # DSL_NVBUF_MEM_TYPE_CUDA_UNIFIED if running on dGPU
        if dsl_info_gpu_type_get(0) == DSL_GPU_TYPE_DISCRETE:

            # If using the new Streammux, then change the memory type of each source
            if dsl_info_use_new_nvstreammux_get():
                retval = dsl_component_nvbuf_mem_type_set_many(
                    ['uri-source-0', 'uri-source-1', 'uri-source-2', 'uri-source-3'],
                    DSL_NVBUF_MEM_TYPE_CUDA_UNIFIED)
                
            # if using the old Streammux we set the memtype of the Streammux itself.    
            else:
                retval = dsl_pipeline_streammux_nvbuf_mem_type_set('pipeline',
                    DSL_NVBUF_MEM_TYPE_CUDA_UNIFIED)
        
        if retval != DSL_RETURN_SUCCESS:
            break
        
        ## Add the listener callback functions defined above
        retval = dsl_pipeline_state_change_listener_add('pipeline', 
            state_change_listener, None)
        if retval != DSL_RETURN_SUCCESS:
            break
        retval = dsl_pipeline_eos_listener_add('pipeline', eos_event_listener, None)
        if retval != DSL_RETURN_SUCCESS:
            break

        # Play the pipeline
        retval = dsl_pipeline_play('pipeline')
        if retval != DSL_RETURN_SUCCESS:
            break

        dsl_main_loop_run()
        retval = DSL_RETURN_SUCCESS
        break

    # Print out the final result
    print(dsl_return_value_to_string(retval))

    dsl_pipeline_delete_all()
    dsl_component_delete_all()

if __name__ == '__main__':
    sys.exit(main(sys.argv))
